pylabrobot/pumps/pump.py

```diff
@@ -1,27 +1,71 @@
-from pylabrobot.machine import MachineFrontend
-
 from .backend import PumpBackend
+from .calibration import PumpCalibration
+from typing import Optional, Union
+from pylabrobot.machine import MachineFrontend
+import asyncio
 
 
 class Pump(MachineFrontend):
-  """ Frontend for a (peristaltic) pump. """
+  """ Frontend for a (peristaltic) pump.
+
+  Attributes:
+      backend: The backend that the pump is controlled through.
+      calibration: The calibration of the pump.
+
+      """
 
-  def __init__(self, backend: PumpBackend):
+  def __init__(self, backend: PumpBackend, calibration: Optional[PumpCalibration] = None):
     self.backend: PumpBackend = backend
+    self.calibration = calibration
 
-  def run_revolutions(self, num_revolutions: float):
+  async def run_revolutions(self, num_revolutions: float):
+    """
+    Run for a given number of revolutions.
+
+    Args:
+        num_revolutions: number of revolutions to run.
+    """
     self.backend.run_revolutions(num_revolutions=num_revolutions)
 
-  def run_continuously(self, speed: float):
-    """ Run continuously at a given speed.
+  async def run_continuously(self, speed: float):
+    """
+    Run continuously at a given speed.
 
     If speed is 0, the pump will be halted.
 
     Args:
-      speed: speed in rpm
+        speed: speed in rpm/pump-specific units.
     """
 
     self.backend.run_continuously(speed=speed)
 
-  def halt(self):
+  async def run_for_duration(self, speed: Union[float, int],
+                             duration: Union[float, int]):
+    """
+    Run the pump at specified speed for the specified duration.
+    Args:
+      speed: speed in rpm/pump-specific units.
+      duration: duration to run pump.
+    """
+    await self.run_continuously(speed=speed)
+    await asyncio.sleep(duration)
+    await self.run_continuously(speed=0)
+
+  async def pump_volume(self, speed: Union[float, int],
+                        volume: Union[float, int]):
+    """
+    Run the pump at specified speed for the specified volume. Note that this function requires
+    the pump to be calibrated at the input speed.
+    Args:
+      speed: speed in rpm/pump-specific units.
+      volume: volume to pump.
+    """
+    if self.calibration is None:
+      raise TypeError(
+        "Pump is not calibrated. Volume based pumping and related functions unavailable.")
+    duration = volume / self.calibration[0]
+    await self.run_for_duration(speed=speed, duration=duration)
```

Comment:
I think running for a certain number of revolutions may be more precise, since pumps could have acceleration / deceleration built in. I know the agrow pumps do not have this level of control. Do you think this should be a backend level command? Do you have other suggestions?

Reply:
I agree, but I think a lot of pumps are not like the Cole Palmer. It would be strongly preferable if it was volume per rotation, as then you won't need the calibration to be speed specific

Reply:
I think we could have it on the backend as an override method.

Reply:
I agree that volume / rotation is more general.

How would you envision implementing it as an override method? The frontend would need to check whether the backend supports this operation, which is not a pattern used elsewhere in plr. I'm open to introducing it here, but would prefer to think about alternatives first.

* We could considering making `pump_volume` a backend-only method, like `run_continuously`. This would involve moving `run_for_duration` out of the frontend, which could get messy.
* Another option is having a parameter `method` (there's probably a better name) with options `"duration"` or `"rotations"`, which determines which backend command we call. What do you think?

Reply:
Also, if the unit of calibration is volume/rotation, won't we need to having something like this:

```py
num_rotations = volume / self.calibration
duration = num_rotations / speed
```

Reply:
I like the option of keeping it on the front end only but checking the backend for the method. I think `calibration_unit` could be the parameter. Would we want to have an attribute that is set on the backend (`backend.calibration_unit`) or do you think another approach would be better?